enterprise/backend/src/metabase_enterprise/semantic_search/embedding.clj

```diff
@@ -19,6 +24,65 @@
    "text-embedding-3-large" 3072
    "text-embedding-ada-002" 1536})
 
+;;; Token Counting for OpenAI Models
+
+(def ^:private ^Encoding openai-encoding
+  "OpenAI tokenizer encoding for cl100k_base (used by text-embedding models)."
+  (delay (.getEncoding (Encodings/newDefaultEncodingRegistry) EncodingType/CL100K_BASE)))
+
+(defn- count-tokens
+  "Count the number of tokens in a text string using OpenAI's cl100k_base encoding."
+  [^String text]
+  (when text
+    (let [^Encoding encoding @openai-encoding]
+      (.size (.encode encoding text)))))
+
+(defn- count-tokens-batch
+  "Count the total number of tokens across multiple text strings."
+  [texts]
+  (reduce + 0 (map count-tokens texts)))
+
+;;; Batching Logic
+
+(def ^:dynamic ^:private *max-tokens-per-batch*
+  "Maximum number of tokens allowed per batch for OpenAI embedding API. The real limit is 8192, but we're using a lower
+  limit in case there are discrepencies between the tokenizers."
+  8000)
+
+(defn- create-batches
```

Comment:
i think this algorithm is separable from the measuring tools `count-tokens` such that you could disentangle the two. makes it super easy to test the function `(defn create-batches [threshold measure coll] ...)` rather than hardcoding `*max-tokens-per-batch*` and `count-tokens`

Reply:
nice suggestion, thanks